src/editor/normalize.ts

```diff
@@ -0,0 +1,135 @@
+import { Editor, Path, Transforms } from "slate";
+import { isListType } from "./helpers";
+import { BlockElement, BlockElementType, SluglineElement } from "./types";
+
+const normalizeVoidSpacer = (editor: Editor, path: Path) => {
+  const prevEntry = Editor.previous(editor, { at: path });
+  const prevElem = prevEntry && (prevEntry[0] as SluglineElement);
+
+  const nextEntry = Editor.next(editor, { at: path });
+  const nextElem = nextEntry && (nextEntry[0] as SluglineElement);
+
+  // if there's a void node before or after the spacer, we don't have to do anything
+  if (
+    prevElem &&
+    Editor.isBlock(editor, prevElem) &&
+    Editor.isVoid(editor, prevElem) &&
+    prevElem.type !== BlockElementType.VoidSpacer
+  ) {
+    return;
+  }
+
+  if (
+    nextElem &&
+    Editor.isBlock(editor, nextElem) &&
+    Editor.isVoid(editor, nextElem) &&
+    nextElem.type !== BlockElementType.VoidSpacer
+  ) {
+    return;
+  }
+
+  // otherwise, remove the node
+  Transforms.removeNodes(editor, { at: path });
+};
+
+const normalizeVoidBlock = (editor: Editor, path: Path) => {
+  const pathRef = Editor.pathRef(editor, path);
+
+  const prevEntry =
+    pathRef.current &&
+    Editor.previous(editor, {
+      at: pathRef.current,
+      mode: "highest",
+    });
+
+  if (!prevEntry) {
+    Transforms.insertNodes(
+      editor,
+      {
+        type: BlockElementType.VoidSpacer,
+        children: [{ text: "" }],
+      },
+      {
+        at: [0],
+        mode: "highest",
+      }
+    );
+    return;
+  } else {
+    const [prevNode, prevPath] = prevEntry;
+    if ((prevNode as SluglineElement).type !== BlockElementType.VoidSpacer) {
+      Transforms.insertNodes(
+        editor,
+        {
+          type: BlockElementType.VoidSpacer,
+          children: [{ text: "" }],
+        },
+        { at: Editor.end(editor, prevPath), mode: "highest" }
+      );
+      return;
+    }
+  }
+
+  const nextEntry =
+    pathRef.current &&
+    Editor.next(editor, { at: pathRef.current, mode: "highest" });
+
+  if (!nextEntry) {
+    // insert a void spacer at the end of the document
+    Transforms.insertNodes(
+      editor,
+      {
+        type: BlockElementType.VoidSpacer,
+        children: [{ text: "" }],
+      },
+      {
+        at: [editor.children.length],
+      }
+    );
+    return;
```

Comment:
Are the returns here and one line 101 necessary?

Reply:
https://github.com/UWmathNEWS/slugline-web/pull/84/commits/47f43914bb06fc8a929d8d02739d83683829f688